xonsh/lexer.py

```diff
@@ -1,13 +1,307 @@
+        state['last'] = n
+    else:
+        yield _new_token('AT', '@', token.start)
+        state['last'] = token
+        for i in handle_token(state, n, stream):
+            yield i
+
+
+def handle_question(state, token, stream):
+    try:
+        n = next(stream)
+    except:
+        n = None
+
+    if n.type == tokenize.ERRORTOKEN and n.string == '?' and \
+            n.start == token.end:
+        yield _new_token('DOUBLE_QUESTION', '??', token.start)
+        state['last'] = n
+    else:
+        yield _new_token('QUESTION', '?', token.start)
+        state['last'] = token
+        for i in handle_token(state, n, stream):
+            yield i
+
+
+def handle_backtick(state, token, stream):
+    try:
+        n = next(stream)
+    except:
+        n = None
+
+    found_match = False
+    sofar = '`'
+    while n is not None:
+        sofar += n.string
+        if n.type == tokenize.ERRORTOKEN and n.string == '`':
+            found_match = True
+            break
+        try:
+            n = next(stream)
+        except:
+            n = None
+    if found_match:
+        yield _new_token('REGEXPATH', sofar, token.start)
+        state['last'] = n
+    else:
+        e = "Could not find matching backtick for regex on line {0}"
+        m = e.format(token.start[0])
+        yield _new_token("ERRORTOKEN", m, token.start)
+
+
+def handle_newline(state, token, stream):
+    try:
+        n = next(stream)
+    except:
+        n = None
+
+    yield _new_token('NEWLINE', '\n', token.start)
+    state['last'] = token
+
+    if n is not None:
+        if n.type != tokenize.ENDMARKER:
+            for i in handle_token(state, n, stream):
+                yield i
+
+
+def handle_lparen(state, token, stream):
+    state['pymode'].append(True)
+    state['last'] = token
+    yield _new_token('LPAREN', '(', token.start)
+
+
+def handle_lbrace(state, token, stream):
+    state['pymode'].append(True)
+    state['last'] = token
+    yield _new_token('LBRACE', '{', token.start)
+
+
+def handle_lbracket(state, token, stream):
+    state['pymode'].append(True)
+    state['last'] = token
+    yield _new_token('LBRACKET', '[', token.start)
+
+
+def handle_rparen(state, token, stream):
+    state['pymode'].pop()
+    state['last'] = token
+    yield _new_token('RPAREN', ')', token.start)
+
+
+def handle_rbrace(state, token, stream):
+    state['pymode'].pop()
+    state['last'] = token
+    yield _new_token('RBRACE', '}', token.start)
+
+
+def handle_rbracket(state, token, stream):
+    state['pymode'].pop()
+    state['last'] = token
+    yield _new_token('RBRACKET', ']', token.start)
+
+
+def handle_error_space(state, token, stream):
+    if not state['pymode'][-1]:
+        state['last'] = token
+        yield _new_token('WS', ' ', token.start)
+    else:
+        yield from []
+
+special_handlers = {
+    tokenize.ENCODING: lambda s, t, st: [],
+    tokenize.COMMENT: lambda s, t, st: [],
+    tokenize.ENDMARKER: lambda s, t, st: [],
```

Comment:
Same lambda is redefined three times.  Probably best to define it once (ie `_empty = lambda s, t, st: []`) and then refer to it multiple times here.